qutebrowser/mainwindow/prompt.py

```diff
@@ -712,6 +728,36 @@ def _init_fileview(self):
         self._file_model.directoryLoaded.connect(
             lambda: self._file_model.sort(0))
 
+    def _get_dirs(self):
+        dirs: Dict[str, List[str]] = {"valid": [], "invalid": []}
+        try:
+            num_rows = self._file_model.rowCount(self._root_index)
+            for row in range(num_rows):
+                index = self._file_model.index(row, 0, self._root_index)
+                hidden = self._to_complete not in index.data()
+                if not hidden:
+                    dirs['valid'].append(index.data())
+                else:
+                    dirs['invalid'].append(index.data())
+        except FileNotFoundError:
+            log.prompt.debug("Directory doesn't exist, can't \
+                             get valid dirs")
+
+        return dirs
+
+    def _do_completion(self, idx, which):
+        filename = self._file_model.fileName(idx)
+        valid_dirs = self._get_dirs()
+        while filename not in valid_dirs['valid'] and idx.isValid():
```

Comment:
Something based on my suggestion earlier seems to work fine for me:

```suggestion
        while idx.isValid() and self._file_view.isIndexHidden(idx):
```

which would then let you simplify the code duplication in `_directories_hide_show_model` again because we don't need `_get_dirs` anymore (as it's only needed in one place).

Reply:
For me the tests would fail when I used this line rather than what I have, but I reduced code duplication between the two.

Reply:
Then let's try to understand what the issue is with the test, rather than ignoring a potential issue and making the code more complex.

It looks like `test_backspacing_path` is failing. The test starts with a directory like `/tmp/pytest-of-florian/pytest-161/test_backspacing_path0/test/foo/`, in the prompt, presses backspace three times to arrive at `.../test_backspacing_path0/test/f` and then expects to get to `foo` again when pressing tab. However, it gets `.../test_backspacing_path0` instead.

pytest keeps its temporary directories around, so you can copy the path and check what's going on for yourself.

So the following is happening:

- With the original code, pressing Tab with `f` entered, qutebrowser jumps to `foo`
- With your code, pressing Tab always jumps to the first directory, and `..` isn't selected at all. This is a bug (like I've told you in IRC already), selecting `..` via tab should be possible.
- With my suggestion, since `..` is the first visible directory, it gets selected when pressing tab.

I'm okay with the behavior changing here a bit - I guess if we filter the visible items down to the matching ones, it's fine if the first tab press goes to `..` instead of going to `foo`. Thus, the only thing you'd have to do is adjust the test to press tab twice.